# tfs-ruby-3-3-3-gnumakefile-in-pass1-msys.patch
#
# Target:  cygwin/GNUmakefile.in (pass1 (msys, toolchain build))
# Applies: ruby 3.3.3 only
# Source:  tebako gem lib/tebako/packager/pass1_patch.rb Pass1MSysPatch#gnumakefile_in_patch_p1
#
# msys pass1: generate export definitions with dlltool (tebako.def),
# compile win32/winmain.c (WINMAINOBJ) and use it for rubyw.exe,
# group-wrap the static-ext ruby.exe link libraries, and comment out the
# default implib/export/program rules.
diff --git a/cygwin/GNUmakefile.in b/cygwin/GNUmakefile.in
index 192a8cc..7e94e90 100644
--- a/cygwin/GNUmakefile.in
+++ b/cygwin/GNUmakefile.in
@@ -46,6 +46,12 @@ SOLIBS := $(DLL_BASE_NAME).res.$(OBJEXT) $(SOLIBS)
 override EXTOBJS += $(if $(filter-out $(RUBYW_INSTALL_NAME),$(@:$(EXEEXT)=)),$(RUBY_INSTALL_NAME),$(@:$(EXEEXT)=)).res.$(OBJEXT)
 RCFILES = $(RUBY_INSTALL_NAME).rc $(RUBYW_INSTALL_NAME).rc $(DLL_BASE_NAME).rc
 RUBYDEF = $(DLL_BASE_NAME).def
+
+# Start of tebako patch
+WINMAINOBJ    = win32/winmain.$(OBJEXT)
+$(WINMAINOBJ): win32/winmain.c
+# End of tebako patch
+
 override LIBRUBY_FOR_LEAKED_GLOBALS := # DLL shows symbols from import library
 
 ruby: $(PROGRAM)
@@ -66,20 +72,21 @@ $(RUBY_EXP) $(LIBRUBY_SO): $(DLL_BASE_NAME).res.$(OBJEXT)
 	  . $(icondirs) $(srcdir)/win32
 
 $(PROGRAM): $(RUBY_INSTALL_NAME).res.$(OBJEXT)
-$(WPROGRAM): $(RUBYW_INSTALL_NAME).res.$(OBJEXT)
+$(WPROGRAM): $(RUBYW_INSTALL_NAME).res.$(OBJEXT) $(WINMAINOBJ)  # tebako patched
 	@rm -f $@
 	$(ECHO) linking $@
 	$(Q) $(PURIFY) $(CC) -mwindows -e $(SYMBOL_PREFIX)mainCRTStartup $(LDFLAGS) $(XLDFLAGS) \
-	  $(MAINOBJ) $(EXTOBJS) $(LIBRUBYARG) $(LIBS) -o $@
+	  $(WINMAINOBJ) $(EXTOBJS) $(LIBRUBYARG) -Wl,--start-group $(LIBS) -Wl,--end-group -o $@  # tebako patched
 $(STUBPROGRAM): $(RUBY_INSTALL_NAME).res.$(OBJEXT)
 
 $(RUBY_EXP): $(LIBRUBY_A)
 	$(ECHO) creating $@
-	$(Q) $(DLLWRAP) \
-	  --output-exp=$(RUBY_EXP) \
-	  --export-all $(LIBRUBY_A) $(LIBS) -o $(PROGRAM)
-	$(Q) $(LDSHARED) $(DLDFLAGS) $(OBJS) dmyext.o $(SOLIBS) -o $(PROGRAM)
-	@rm -f $(PROGRAM)
+	$(Q) dlltool --output-lib=$(LIBRUBY) --output-def=tebako.def --export-all $(LIBRUBY_A) --output-exp=$(RUBY_EXP)   # tebako patched
+
+	  # tebako patched --output-exp=$(RUBY_EXP) \
+	  # tebako patched --export-all $(LIBRUBY_A) $(LIBS) -o $(PROGRAM)
+# tebako patched  $(Q) $(LDSHARED) $(DLDFLAGS) $(OBJS) dmyext.o $(SOLIBS) -o $(PROGRAM)
+	# tebako patched @rm -f $(PROGRAM)
 
 GNUmakefile:	$(srcdir)/cygwin/GNUmakefile.in
 
